text-generation-inference/server/text_generation_server/generator.py

```diff
@@ -646,3 +599,170 @@ def from_pretrained(
         return cls(model, tokenizer)
+
+
+class GeneratorCommand(Enum):
+    INFO = 0
+    WARMUP = 1
+    PREFILL = 2
+    DECODE = 3
+    FILTER = 4
+    CLEAR = 5
+    DELETE = -1
+
+
+def _mp_fn(rank,
+           model_path: str,
+           revision: str,
+           max_batch_size: int,
+           max_sequence_length: int,
+           root_mailbox: RootMailbox):
+    device = xm.xla_device()
+    world_size = xm.xrt_world_size()
+    # create agent mailbox out of root's one
+    mailbox = AgentMailbox(root_mailbox)
+
+    logger.debug(
+        f"Rank {rank} on {device} real device {xm.xla_real_devices([device])} ordinal {xm.get_ordinal()} "
+        + f"world size {world_size}"
+    )
+
+    generator = TpuGeneratorSingleThread.from_pretrained(model_path, revision, max_batch_size, max_sequence_length)
+    # TODO: maybe model_config can be removed from mailbox
+
+    def return_to_caller(*data):
+        # consider adding a rendezvous here
+        if rank == 0:
+            xm.mark_step()
+            mailbox.send(*data)
+
+
+    while True:
+        xm.rendezvous("start")
+        if rank == 0:
+            mailbox.agent_ready.set()
+            mailbox.receive()
+        # Wait for rank 0 to receive command
+        xm.rendezvous("wait_command")
+        command, data = mailbox.command_data
+        logger.debug(f"Generator@{rank} {command.name}")
+        if command == GeneratorCommand.INFO:
+            info = generator.info
+            return_to_caller(info.SerializeToString())
+        if command == GeneratorCommand.WARMUP:
+            batch = Batch.FromString(data[0])
+            return_to_caller(generator.warmup(batch=batch))
+        if command == GeneratorCommand.PREFILL:
+            batch = Batch.FromString(data[0])
+            generations, cached_batch = generator.prefill(batch=batch)
+            return_to_caller([g.SerializeToString() for g in generations], cached_batch.SerializeToString())
+        if command == GeneratorCommand.DECODE:
+            batches = [CachedBatch.FromString(b) for b in data[0]]
+            generations, cached_batch = generator.decode(batches=batches)
+            s_cached_batch = cached_batch.SerializeToString() if cached_batch is not None else None
+            return_to_caller([g.SerializeToString() for g in generations], s_cached_batch)
+        if command == GeneratorCommand.FILTER:
+            batch_id, request_ids = data
+            cached_batch = generator.filter(batch_id, request_ids)
+            return_to_caller(cached_batch.SerializeToString())
+        if command == GeneratorCommand.CLEAR:
+            generator.clear()
+        if command == GeneratorCommand.DELETE:
+            if rank == 0:
+                # Set agent to ready
+                mailbox.agent_ready.set()
+            break
+
+
+def model_loop_fn(*args):
+    """Spawn processes in the TPUs forwarding arguments"""
+    xmp.spawn(_mp_fn, args=(args), join=True, daemon=False)
+
+class TpuGenerator(Generator):
```

Comment:
Let's maybe suggest it's something involving mp? `TpuGeneratorShared`?

Reply:
Umh, my idea was to leave it as `TpuGenerator`, so code using it will be left unchanged, and also because I think we will always spawn on all available TPUs. I can rename if though if you prefer.